### DISCUSSION TITLE
What is the expected lifetime of `cfg` and `svc` instances?

### DISCUSSION THREAD

**Discussion by mattjohnsonpint** [aws/aws-sdk-go-v2#2566] [Q&A] [answered]
Hi.  I have a general question about config and service instances. Consider the initialization code, as seen in many examples:

```go
cfg, err := config.LoadDefaultConfig(ctx)
if err != nil {
    log.Fatalf("unable to load SDK config, %v", err)
}
svc := s3.NewFromConfig(cfg)
```

My application will list and retrieve objects from S3 _frequently_.  Should create a new `cfg` and `svc` instance with each call?  Or should I create them on application startup and reuse those instances for every call?

It _seems_ to work either way, but I'm not sure about:
- Is the cost of creating them is small or large? Is it even worth reusing them?
- Are there any thread safety issues with multiple goroutines using the same `svc` instance?
- Does either the `cfg` or `svc` instance keep any tokens that might expire?

Any other thoughts would be useful, and I would hope the docs would be updated to explain this better also. Thanks.

Update: I did see [this doc](https://aws.github.io/aws-sdk-go-v2/docs/making-requests/#concurrently-using-service-clients), which seems to indicate reusing service clients is thread-safe.  But it doesn't really answer my other concerns.  I'm mostly concerned about whether it's safe to keep the service client for a long period of time (days, weeks, etc.).  Will it refresh its own authentication automatically?  And is this advised?  Thanks.

**Comment by lucix-aws** [comment, marked as answer]
> Is the cost of creating them is small or large? Is it even worth reusing them?

I would classify service clients as rather lightweight in this SDK. That said you do generally want to keep client instances around simply because of credential provision. If you're on a platform like EC2 or EKS, where credential provision requires network calls, you can save repeated instances of those calls because the client will cache credentials until they expire. This is not a _hard_ rule though, just know that this is something to consider.

> Are there any thread safety issues with multiple goroutines using the same svc instance?

No - we explicitly designate clients to be goroutine-safe

> Does either the cfg or svc instance keep any tokens that might expire?

Just credentials. The SDK automatically refreshes credentials _where dynamic refresh is possible_. Dynamic refresh is possible for all sources minus environment and shared config. In general, if you're deploying software to a compute environment where long-running applications are possible (e.g. EC2) credential refresh is automatic.

**Reply by mattjohnsonpint** [reply]
Thanks!
